homeassistant/components/mqtt/mixins.py

```diff
@@ -1115,14 +1121,47 @@ async def async_publish(
     def config_schema() -> vol.Schema:
         """Return the config schema."""
 
+    def _set_entity_name(self, config: ConfigType) -> None:
+        """Help setting the entity name if needed."""
+        entity_name: str | None | UndefinedType = config.get(CONF_NAME, UNDEFINED)
+        # Only set _attr_name if it is needed
+        if entity_name is not UNDEFINED:
+            self._attr_name = entity_name
+        elif not (use_device_class_name := self._default_to_device_class_name()):
+            # Assign the default name
+            self._attr_name = self._default_name
+        elif use_device_class_name:
+            # Follow name of device class
+            self._attr_has_entity_name = True
+            if CONF_DEVICE in config and CONF_NAME not in config[CONF_DEVICE]:
+                _LOGGER.info(
+                    "No device name set in config: %s,"
+                    "'has_entity_name' was set to True, assuming the entity is linked "
+                    "to an existing shared device, see also "
+                    "https://developers.home-assistant.io/docs/core/entity/#entity-naming",
+                    config,
+                )
+            return
+        if CONF_DEVICE in config:
+            self._attr_has_entity_name = device_named = CONF_NAME in config[CONF_DEVICE]
```

Comment:
```suggestion
            self._attr_has_entity_name = True
```

Reply:
I needed to update some tests for this to be set to True, as this influences the default entity name to gecome a `none_` prefix when no device name is set.